### ISSUE TITLE
ChainCache save value with infinite lifetime when propagate to fastest cache

### ISSUE THREAD

**Issue by lchenay** [doctrine/cache#246]
In case of item found in the chain of cache, a propagation of the value is made backward.
Those save are made without any TTL : https://github.com/doctrine/cache/blob/2b0e130afa0d70bafaf3da6c6d2022649562bc97/lib/Doctrine/Common/Cache/ChainCache.php#L50-L53
It lead to have infinite lifetime on those fastest cache ; and so many issue in production.

Here a test case to reproduce:
```php
    public function testFetchPropagateToFastestCacheWithTTL()
    {
        $cache1 = new ArrayCache();
        $cache2 = new ArrayCache();

        $cache2->save('bar', 'value', 1);

        $chainCache = new ChainCache([$cache1, $cache2]);

        self::assertFalse($cache1->contains('bar'));

        $result = $chainCache->fetch('bar');

        self::assertEquals('value', $result);
        self::assertTrue($cache1->contains('bar'));

        sleep(2);

        self::assertFalse($cache2->contains('bar'));
        self::assertFalse($cache1->contains('bar'));
        self::assertFalse($chainCache->contains('bar'));
    }
```

2 proposals that i see:
* Maybe we could have a short default ttl inside ChainCache that will be used in case of chain propagation?
* Get the remained ttl and propagate with this value ? (currently ttl is not exposed)

**Comment by lchenay** [comment]
For proposal 2, many backend don't have a getter to retrieve the remain ttl.
A quick solution could be to wrap the cachedData in an object containing the {cachedData, ttl}.
But it will prevent to reuse the same Cache server without instancing a ChainCache as the stored data is not in the same format.

Without a new elegant solution on proposal 2, not sure it's worth it.

**Comment by holtkamp** [comment]
mmm, just encountered this issue and it is actually quite "severe" right? It might also explain some crazy behaviour I have encountered for which I usually resorted to using a "TTL-specific" cacheEntryIdentifier...

**Description**
Mainly for my own understanding, a short description with an example.

Imagine an application running somewhere in the cloud, which [typically reboots once every 24 hours](https://devcenter.heroku.com/articles/dynos#restarting) with the following ChainCache:

- fast and volatile: is cleared on each server restart
- slow and persistant

Now suppose :
- a cache entry X is saved to the ChainCache with a TTL of 4 hours
  - cache entry X is saved to the fast cache with a TTL of 4 hours
  - cache entry X to the slow cache with a TTL of 4 hours
- the server is restarted, clearing the fast and volatile cache (APCu)
- cache entry X is looked up
  - cache entry X is _not_ found in the fast (APCu) cache
  - cache entry X is found in the slow cache
  - cache entry X will be propagated to / stored in the fast cache with _infinite_ TTL
- after 4 hours, cache entry X is looked up again
  - cache entry X is found in the fast (APCu) cache and used, potentially leading to strange behavior
- ... repeat, repeat, repeat using invalid cache entry until next server restart
 

**Possible solutions**
As suggested here https://github.com/doctrine/cache/issues/246#issue-300591229, the second proposal of trying to determine the remaining TTL would be the most elegant, but quite challenging to implement for all various CacheProviders. Maybe we can check whether a CacheProvider implements a specific interface like:

```php
interface RemainingTimeToLiveAwareCacheProvider
{
	public function getRemainingTimeToLiveInSeconds(string $cacheEntry): int
}
```

In case a CacheProvider does not implement such interface, we can fall back to solution number 1: allow to define a default TTL for the ChainCache, which defaults to null, but can be set to a relative short periods to prevent problems / strange behaviour, but does increase the load on the slow cache...

**Comment by holtkamp** [comment]
Anyone got feedback on the suggested approach?

**Comment by holtkamp** [comment]
@Ocramius this issue:

- was fixed in https://github.com/doctrine/cache/pull/335
- was released in https://github.com/doctrine/cache/releases/tag/1.10.1
  - without mentioning it in the "release notes"
- and can be closed

**Comment by malarzm** [comment]
Thanks for the heads up!

**Comment by holtkamp** [comment]
@malarzm sorry, I misread this page:

https://github.com/doctrine/cache/compare/1.10.1...master

`1.10.1` was tagged on 27-05-2020 and https://github.com/doctrine/cache/pull/335 was merged one day later 28-05-2020

So I think this issue needs to be re-opened until new release is tagged? Probably a patch version `1.10.2` suffices?

**Comment by holtkamp** [comment]
@beberlei can you have a look at this? Been trying to solve quite a serious bug since https://github.com/doctrine/cache/issues/246#issuecomment-418731044, only think remaining is tagging it in a new release...

**Comment by malarzm** [comment]
I'll try to backport your fix to 1.10.x branch and cut a new release this evening :)
